node/resolvers/directives/auditAccess.ts

```diff
@@ -0,0 +1,63 @@
+import type { GraphQLField } from 'graphql'
+import { defaultFieldResolver } from 'graphql'
+import { SchemaDirectiveVisitor } from 'graphql-tools'
+
+import sendAuthMetric, { AuthMetric } from '../../utils/metrics/auth'
+import { getCheckUserPermission } from './withPermissions'
+
+export class AuditAccess extends SchemaDirectiveVisitor {
+  public visitFieldDefinition(field: GraphQLField<any, any>) {
+    const { resolve = defaultFieldResolver } = field
+
+    field.resolve = async (
+      root: any,
+      args: any,
+      context: Context,
+      info: any
+    ) => {
+      const {
+        clients: { storefrontPermissions },
+        vtex: { adminUserAuthToken, storeUserAuthToken, account, logger },
+        request,
+      } = context
+
+      const operation = field.astNode?.name?.value ?? request.url
+      const forwardedHost = request.headers['x-forwarded-host'] as string
+      const caller = request.headers['x-vtex-caller'] as string
+
+      const hasAdminToken = !!(
+        adminUserAuthToken ?? (context?.headers.vtexidclientautcookie as string)
+      )
+
+      const hasStoreToken = !!storeUserAuthToken
+      const hasApiToken = !!request.headers['vtex-api-apptoken']
+
+      let role
+      let permissions = []
+
+      if (hasAdminToken || hasStoreToken) {
+        const checkUserPermissions = await getCheckUserPermission(
```

Comment:
Same here as I commented for the [other similar PR](https://github.com/vtex-apps/storefront-permissions/pull/118#discussion_r1377481834)

Reply:
@Rudge this one is the only one still pending an update, to avoid awaiting within the resolver function.

Reply:
Sorry, could you look again, please?